submariner/IPV6-datapath.md

```diff
@@ -0,0 +1,129 @@
+* Pods (including host-networking) to Pods.
+
+## Proposal
+
+Currently, Submariner fully supports IPV4 inter-cluster connectivity, including egress in-cluster routing to reach GW node, GlobalNet, and inter-cluster tunnels.
+
+The idea is to duplicate intra-cluster and inter-cluster connectivity components also for IPV6.
+
+The active Gateway Engine communicates with the central Broker to advertise its Endpoint and Cluster resources to the other clusters connected to the Broker, also ensuring that it is the sole Endpoint for its cluster.
+The Endpoint resource fields should include IP addresses according to the cluster’s networking configuration.
+For example for a dual-stack cluster HealthCheckIP, PrivateIP, PublicIP and Subnets should consist of both IPv4 and IPV6 addresses.  
+
+The Route Agent running in the cluster learns about the local Endpoint and remote Endpoints and setup the necessary V4,V6 infrastructure to route cross-cluster traffic from all nodes to the active Gateway Engine node. 
+
+The active Gateway Engine also establishes a watch on the Broker to learn about the active Endpoint and Cluster resources advertised by the other clusters. Once two clusters are aware of each other’s Endpoints, they can establish a secure tunnel/s based on the content of the remote Endpoint and local Endpoint through which traffic can be routed. A tunnel should be created only if the local Endpoint networking type matches remote Endpoint ipfamily.
+
+The next diagram illustrates Submariner’s datapath architecture for kube-proxy based CNIs:
+![non-ovnk-architecture](./images/dual-stack-arch-for-non-ovnk.png)
+
+With the proposed architecture, Submariner needs to establish both V4 and V6 intra-cluster egress routing to the GW node in the case of dual-stack.
+
+Pod IPV4 egress packets for CNI != OVNK and cable-driver=libreswan will be:
+![non-ovnk-ipv4-egress](./images/ipv4-non-ovnk-egress-packets.png)
+
+And Pod IPV6 egress packets for the same configuration will be:
+![non-ovnk-ipv6-egress](./images/ipv6-non-ovnk-egress-packets.png)
+
+**Note**: In future, we may optimize this architecture for a dual-stack case.
+For example by using only the intra-cluster V4 VxLAN to route V4 and V6 traffic to the GW.
+
+## Datapath breakdown
+
+### Gateway
+
+To support IPV6 the gateway should:
+
+* discover publicIP, privateIP, healthcheckIP and cluster’s subnets for each IP family.  
+  * Note: gateway should address corner cases related to this change, for example in dual-stack environment only the V4 publicIP address is successfully resolved.    
+* run NAT Discovery per IP family in remote Endpoint.
+* advertise in local Endpoint IP details based on cluster networking type, for example in duals-stack cluster both V4 Public IP and V6 Public IP should be advertised in Endpoint.   
+* continue advertising a **single** Endpoint. in case of a dual-stack cluster, fields should consist of both V4 and V6 addresses separated.
```

Comment:
I think we should change the appropriate IP fields to arrays and bump the API version and implement a webhook converter so each version can be served for backwards compatibility.

Reply:
If we are changing any CRDs, we should also consider the impact during upgrades as there will be brownfield deployments.

Reply:
yes - a webhook converter would handle that, ie convert to and from both versions depending on client requests.

Reply:
In that case, I guess we need to be persistent and change also other resources , for example 
GlobalnetCIDRRange and ClustersetIPCIDRRange in Broker resource. right ?

Reply:
yeah whatever is needed

Reply:
@tpantelis Why not follow the core API example add a new field with `xxxIPs` which is an array? This makes it easy to handle upgrades and allows us to depricate old `xxxIP` field over time.

Reply:
We could do that but then we’d have to handle both fields in various places in code going forward. A webhook converter handles all that transparently.  Webhooks and CRD versioning were designed for cases like this.  K8s core API adds new fields and deprecates old ones b/c they’re  not CRDd and much harder to bump versions.  Web hooks can discuss further (next year 🙂)